clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -198,34 +198,30 @@ protected abstract void onJoinComplete(int generation,
                                            ByteBuffer memberAssignment);
 
     /**
-     * Block until the coordinator for this group is known and is ready to receive requests.
-     */
-    public synchronized void ensureCoordinatorReady() {
-        // Using zero as current time since timeout is effectively infinite
-        ensureCoordinatorReady(0, Long.MAX_VALUE);
-    }
-
-    /**
+     * Visible for testing.
+     * 
      * Ensure that the coordinator is ready to receive requests.
-     * @param startTimeMs Current time in milliseconds
-     * @param timeoutMs Maximum time to wait to discover the coordinator
+     *
+     * @param timeoutMs   Maximum time to wait to discover the coordinator
      * @return true If coordinator discovery and initial connection succeeded, false otherwise
      */
-    protected synchronized boolean ensureCoordinatorReady(long startTimeMs, long timeoutMs) {
-        long remainingMs = timeoutMs;
+    public synchronized boolean ensureCoordinatorReady(final long timeoutMs) {
```

Comment:
That makes sense. I was thinking that you'd have the same number of calls (either on the caller or on the callee's side), but that's the case only if you're trying to be super accurate. I think the deadline style you proposed would eliminate this problem.

I guess an equivalent to the deadline approach is to pass in (start,timeout) everywhere, but instead of re-computing start and timeout for each inner call, just using the start and timeout of the outermost request. This approach didn't occur to me before. Deadline is simpler still, though, since it's a scalar value and contains the same information.

Reply:
Regarding the visibility change: There was another method `public synchronized void ensureCoordinatorReady()` that delegated to this one. It needed to be public because the WorkerCoordinator in Streams wants to expose it for a test. I got rid of that one to force all callers to specify a timeout.

I can fix it by adding a package-private delegating method in WorkerCoordinator to expose the method for its test.